webapp/channels/src/components/gif.tsx

```diff
@@ -0,0 +1,176 @@
+    if (typeof canvas.getContext !== 'function') {
+        return null;
+    }
+
+    canvas.width = img.width;
+    canvas.height = img.height;
+    const ctx = canvas.getContext('2d');
+    ctx && ctx.drawImage(img, 0, 0);
+    return canvas.toDataURL();
+};
+
+const calcDurationGif = (file: any) => {
+    return new Promise((resolve, reject) => {
+        try {
+            const arr = new Uint8Array(file);
+            let duration = 0;
+            for (let i = 0; i < arr.length; i++) {
+                if (arr[i] === 0x21 &&
+                arr[i + 1] === 0xF9 &&
+                arr[i + 2] === 0x04 &&
+                arr[i + 7] === 0x00) {
+                    const delay = (arr[i + 5] << 8) | (arr[i + 4] & 0xFF);
+                    duration += delay < 2 ? 10 : delay;
+                }
+            }
+            resolve(duration / 100);
+        } catch (e) {
+            reject(e);
+        }
+    });
+};
+
+type Props = {
+    src: string;
+    autoplay: boolean;
+    handleLoad: () => void;
+    handleError: () => void;
+    onClick: () => void;
+    onKeyDown: () => void;
+}
+
+const GifPlayerContainer = (props: Props) => {
+    const {autoplay, src, handleLoad, handleError, onClick, onKeyDown} = props;
+
+    const [playing, setPlaying] = useState(autoplay);
+    const [providedGif] = useState(src);
+    const [actualGif] = useState(src);
+    const [actualStill, setActualStill] = useState('');
+    const [gifLoopCount, setGifLoopCount] = useState(1);
+    const intervalId = useRef<any>();
+    const gif = document.getElementById('gif') as HTMLImageElement;
+    let updateId = -1;
+
+    useEffect(() => {
+        const updateImages = () => {
+            if (providedGif) {
+                const updatedId = ++updateId;
+                preload(providedGif, (img: HTMLImageElement) => {
+                    if (updateId === updatedId) {
+                        const still = firstGifFrameUrl(img);
+                        if (still) {
+                            setActualStill(still);
+                        }
+                    }
+                });
+            }
+        };
+
+        updateImages();
+    }, [providedGif, updateId]);
+
+    useEffect(() => {
+        if (gifLoopCount === 6) {
+            setPlaying(false);
+            setGifLoopCount(0);
+            clearInterval(intervalId.current);
+        }
+    }, [gifLoopCount]);
+
+    useEffect(() => {
+        const countGifAnimation = () => {
+            gif && gifLoopCounterInit(gif.src);
+        };
+
+        countGifAnimation();
+    }, [gif]);
+
+    const gifLoopCounterInit = useCallback((fileSrc: string) => {
+        const request = new XMLHttpRequest();
+        request.open('GET', fileSrc, true);
+        request.responseType = 'arraybuffer';
+        request.addEventListener('load', async () => {
+            calcDurationGif(request.response).then((duration: any) => {
+                let newCount = gifLoopCount;
```

Comment:
This function is created once, with one particular value of `gifLoopCount` (1). Therefore the `newCount` will never grow.